wallet: warn against accidental unsafe older() import

Sjors opened the pull request:
[BIP 379](https://github.com/bitcoin/bips/blob/master/bip-0379.md) ([Miniscript](https://bitcoin.sipa.be/miniscript/)) allows relative height and time locks that have no consensus meaning in [BIP 68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki) (relative timelocks) / [BIP 112](https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki) (`CHECKSEQUENCEVERIFY`). This is (ab)used by some protocols, e.g. [by Lightning to encode extra data](https://delvingbitcoin.org/t/exploring-extended-relative-timelocks/1818/23), but is unsafe when used unintentionally: `older(65536)` is equivalent to `older(1)`.

This PR emits a warning when `importdescriptors` contains such a descriptor.

The first commit makes `SEQUENCE_LOCKTIME` flags reusable by other tests.

The main commit adds the `ForEachNode` helper to `miniscript.h` which is then used in the `MiniscriptDescriptor` constructor to check for `Fragment::OLDER` with unsafe values. These are stored in `m_warnings`, which the RPC code then collects via `Warnings()`.

It adds both a unit and functional test.

---

A previous version of this PR prevented the import, unless the user opted in with an `unsafe` flag. It also used string parsing in the RPC code.

---

Based on:
- [x] https://github.com/bitcoin/bitcoin/pull/33914

Sjors committed:
test: move SEQUENCE_LOCKTIME flags to script

Sjors commented on src/wallet/rpc/backup.cpp:
```diff
@@ -158,6 +160,30 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
+
+        // Additional safety checks for valid descriptors
+        size_t pos{0};
+        while (!unsafe && (pos = descriptor.find("older(", pos)) != std::string::npos) {
```
Maybe, but I'm worried it would add a ton of code churn in order to either:
1. bubble these values up; or
2. have the descriptor and miniscript parser take an `unsafe` argument to change its behavior (or some more general `parse_options` argument)

brunoerg reviewed (commented):
I've reviewed the code and it looks great - Just a question: Instead of having this new field and prevent the import (any user have reported an accidental import?), couldn't we just throw a warning?

brunoerg reviewed (approved):
ACK fe06b92429206aee66d33c843195235f4bb27181 - This new approach of just warning is better than the previous one, nice! I haven't reviewed the code in depth, just did a light code review but tested the behavior in practice and worked fine. 

Also, I ran a mutation analysis on these changes and the only uncaught mutant is:
```diff
         // Traverse miniscript tree for unsafe use of older()
         miniscript::ForEachNode(*m_node, [&](const miniscript::Node<uint32_t>& node) {
-            if (node.fragment == miniscript::Fragment::OLDER) {
+            if (1==1) {
                 const uint32_t raw = node.k;
                 const uint32_t value_part = raw & ~CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG;
                 if (value_part > CTxIn::SEQUENCE_LOCKTIME_MASK) {
```

it means we're not really testing that the warning is produced exclusively for the `older()` fragment, nothing critical and can be ignored but I think a simple test case could solve it. e.g.:

```diff
diff --git a/src/test/descriptor_tests.cpp b/src/test/descriptor_tests.cpp
index 9e06e78fe9..fb35dfd05a 100644
--- a/src/test/descriptor_tests.cpp
+++ b/src/test/descriptor_tests.cpp
@@ -1260,6 +1260,15 @@ BOOST_AUTO_TEST_CASE(descriptor_test)

 BOOST_AUTO_TEST_CASE(descriptor_older_warnings)
 {
+    {
+        FlatSigningProvider keys;
+        std::string err;
+        // Using after() with a large timestamp (> 65535)
+        auto descs = Parse("wsh(and_v(v:pk(0379e45b3cf75f9c5f9befd8e9506fb962f6a9d185ac87001ec44a8d3df8d4a9e3),after(1000000)))", keys, err, /*require_checksum=*/false);
+        BOOST_REQUIRE(!descs.empty());
+        BOOST_CHECK(descs[0]->Warnings().empty()); // Ensure the older() warning is really being applied only for this fragment
+    }
+
```

Sjors commented:
@brunoerg thanks, I added your test against false positive warnings for `after()`, which passes.

brunoerg reviewed (approved):
reACK 5e9737ff493a96566f4fc11b1733b4b5a5393f6b

Sjors commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -58,6 +59,7 @@ def test_importdesc(self, req, success, error_code=None, error_message=None, war
         if 'warnings' in result[0]:
             observed_warnings = result[0]['warnings']
         assert_equal("\n".join(sorted(warnings)), "\n".join(sorted(observed_warnings)))
+        self.log.debug(result)
```
Yes, it's only printed when you set the log level to debug and it's useful for debugging if the test breaks.

pythcoiner commented:
tACK 5e9737ff49 code review + test importing few "Liana" descriptor with timelocks > 65535

pythcoiner commented on src/script/miniscript.h:
```diff
@@ -7,8 +7,10 @@
 
 #include <algorithm>
 #include <compare>
+#include <concepts>
```
build success even w/o those 2 (added) includes on my side

Sjors replied:
That doesn't really say much, because indirect includes will succeed but we try to avoid those. Though maybe not for std stuff?

Sjors replied:
The Tidy CI job runs IWYU and doesn't complain: https://github.com/bitcoin/bitcoin/actions/runs/17605378583/job/50015040929?pr=33135#step:9:6218

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1530,7 +1542,24 @@ class MiniscriptDescriptor final : public DescriptorImpl
+                if (value_part > CTxIn::SEQUENCE_LOCKTIME_MASK) {
+                    const bool is_time = (raw & CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG) != 0;
+                    if (is_time) {
+                        m_warnings.push_back(strprintf("older(%u - 1<<22) > 65535 is unsafe", raw));
```
In 5e9737ff493a96566f4fc11b1733b4b5a5393f6b "wallet: warn against accidental unsafe older() import"

This warning is basically incomprehensible. It would be better to actually write out what it means, i.e. "time based relative locktime > 65535 is unsafe"

Sjors replied:
I went for a slight variant.

scgbckbone reviewed (commented):
What is the reason for allowing to import `older(65536)` ? I do not see rationale here, only mention of "previous version" that disallowed importing. I'd prefer older version if that is the case. Complete prohibition, even with `unsafe`. Importing `older(65536)` is not what user wants if resulting script is `older(1)` anyways.

scgbckbone commented:
Thanks Rob. I completely forget about that 22nd bit to change to time-based lock.

To update my previous comment: anything without consensus meaning should not be allowed to import. 

Only `older(1 - 65535)` & `older(4194305 - 4259839)` should be importable.

Sjors commented:
As I mentioned in the PR description, there are applications - including the lightning protocol - that use these extra bits. I don't know if they use descriptors internally to represent them, but if they do, then banning these values would mean they can't import these descriptors.

There's also nothing in the BIP that bans (or even discourages or warns about) these values, so we wouldn't be compatible with other wallets.

pythcoiner commented:
reACK [fb72cc3](https://github.com/bitcoin/bitcoin/pull/33135/commits/fb72cc33be570d12fc6b76146fc89047e58f5aaf)

brunoerg reviewed (approved):
reACK fb72cc33be570d12fc6b76146fc89047e58f5aaf

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1537,7 +1549,24 @@ class MiniscriptDescriptor final : public DescriptorImpl
+                        m_warnings.push_back(strprintf("relative locktime > 65535 * 512 seconds is unsafe"));
+                    } else {
+                        m_warnings.push_back(strprintf("older(%u) > 65535 is unsafe", raw));
+                    }
```
Nit: some consistency in the warning format would be easier on the eyes.

```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index 8289120164..9b3f73466a 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -1557,11 +1557,11 @@ public:
                 const uint32_t raw = node.k;
                 const uint32_t value_part = raw & ~CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG;
                 if (value_part > CTxIn::SEQUENCE_LOCKTIME_MASK) {
-                    const bool is_time = (raw & CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG) != 0;
-                    if (is_time) {
-                        m_warnings.push_back(strprintf("relative locktime > 65535 * 512 seconds is unsafe"));
+                    const bool is_time_based = (raw & CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG) != 0;
+                    if (is_time_based) {
+                        m_warnings.push_back(strprintf("time-based relative locktime: older(%u) > (65535 * 512) seconds is unsafe"));
                     } else {
-                        m_warnings.push_back(strprintf("older(%u) > 65535 is unsafe", raw));
+                        m_warnings.push_back(strprintf("height-based relative locktime: older(%u) > 65535 is unsafe", raw));
                     }
                 }
             }
```

rkrux reviewed (approved):
lgtm ACK 189437a3dedefce3945ab2a8c05785eb283c2c6f

Shared a non-blocking nit.

brunoerg reviewed (approved):
reACK 189437a3dedefce3945ab2a8c05785eb283c2c6f

Sjors committed:
wallet: warn against accidental unsafe older() import

BIP 379 allows height and time locks that have no consensus meaning in BIP 68 / BIP 112.
This is used by some protocols like Lightning to encode extra data, but is unsafe when
used unintentionally. E.g. older(65536) is equivalent to older(1).

This commit emits a warning when importing such a descriptor.

It introduces a helper ForEachNode to traverse all miniscript nodes.

rkrux reviewed (approved):
lgtm re-ACK 76c092ff805833a9adf84f669f0455bc2e0bba8b

Sjors commented:
@brunoerg @pythcoiner can you re-review the latest push?

brunoerg reviewed (approved):
reACK 76c092ff805833a9adf84f669f0455bc2e0bba8b
